domain/machine/state/machine_cloud_instance.go

```diff
@@ -148,3 +149,51 @@ WHERE machine_uuid=$instanceTag.machine_uuid
 		return nil
 	})
 }
+
+// InstanceId returns the provider specific instance id for this machine.
+// If the machine is not provisioned, it returns a NotProvisionedError.
+func (st *State) InstanceId(ctx context.Context, machineId string) (string, error) {
+	db, err := st.DB()
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+
+	machineIDParam := sqlair.M{"machine_id": machineId}
+	query := `
+SELECT &M.instance_id
+FROM machine AS m
+		JOIN machine_cloud_instance AS mci ON m.uuid = mci.machine_uuid
+WHERE m.machine_id = $M.machine_id;
+`
+	queryStmt, err := st.Prepare(query, sqlair.M{})
```

Comment:
We don't want to be using maps (`sqlair.M`) unless we have no option. We should use types.

```go
type MachineID struct {
    ID string `db:"machine_id"`
}
```